### PR TITLE
Add `regex_program` searching APIs and related java classes

### PR DESCRIPTION
**Pull request by cindyyuanjiang**
Signed-off-by: Cindy Jiang <cindyj@nvidia.com>

## Description
This PR adds [contains_re, matches_re, count_re, findall](https://docs.rapids.ai/api/libcudf/nightly/group__strings__contains.html) related `regex_program` java APIs and unit tests, and `RegexProgram`, `RegexFlag`, and `CaptureGroups` java classes to support these `regex_program` APIs.

Part of work for https://github.com/NVIDIA/spark-rapids/issues/7295.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by firestarman — APPROVED** [review]

**Review by jlowe — COMMENTED** [review]
The new classes seem OK, but normally we do not add classes and APIs without the ability for the user to meaningfully wield it.  There are no tests of these new APIs to prove they work and are useful, and there's nothing that can be meaningfully done with them as-is.

Should we at least add this API to some Table or ColumnView method to both provide a test surface for it and a meaningful way to start using it?  We wouldn't have to put the new class in all potential places in one PR, but not putting it anywhere seems a bit less than ideal.

**Comment by davidwendt** [comment]
Is there a way to test this through the Spark test suite before merging so we can catch any potential errors like those found during the end of the last release?

**Comment by cindyyuanjiang** [comment]
> The new classes seem OK, but normally we do not add classes and APIs without the ability for the user to meaningfully wield it. There are no tests of these new APIs to prove they work and are useful, and there's nothing that can be meaningfully done with them as-is.

Yes, thank you. I will update to include some usage of the added classes.

**Comment by cindyyuanjiang** [comment]
> Is there a way to test this through the Spark test suite before merging so we can catch any potential errors like those found during the end of the last release?

Yes, I will test the changes in the Spark plugin to avoid issues encountered before. Thanks!

File: java/src/main/java/ai/rapids/cudf/ColumnView.java [hunk 1]
```diff
@@ -3187,11 +3209,33 @@ public final ColumnVector matchesRe(String pattern) {
...
+    return containsRe(new RegexProgram(pattern, CaptureGroups.NON_CAPTURE));
+  }
+
+  /**
+   * Returns a boolean ColumnVector identifying rows which
+   * match the given RegexProgram pattern starting at any location.
+   *
+   * ```
+   * cv = ["abc","123","def456"]
+   * result = cv.contains_re("\\d+")
```
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ColumnView.java:3223]: This example still uses the deprecated method, needs to be updated to show how to use the new API.

File: java/src/main/java/ai/rapids/cudf/ColumnView.java [hunk 2]
```diff
@@ -3164,11 +3164,33 @@ public final ColumnVector clamp(Scalar lo, Scalar loReplace, Scalar hi, Scalar h
...
+    return matchesRe(new RegexProgram(pattern, CaptureGroups.NON_CAPTURE));
+  }
+
+  /**
+   * Returns a boolean ColumnVector identifying rows which
+   * match the given regex program pattern but only at the beginning of the string.
+   *
+   * ```
+   * cv = ["abc","123","def456"]
+   * result = cv.matches_re("\\d+")
```
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ColumnView.java:3178]: This example still uses the deprecated method, needs to be updated to show how to use the new API.

Comment (cindyyuanjiang) [reply java/src/main/java/ai/rapids/cudf/ColumnView.java]: Thank you! Changes updated.

File: java/src/main/java/ai/rapids/cudf/ColumnView.java [hunk 3]
```diff
@@ -3164,20 +3164,43 @@ public final ColumnVector clamp(Scalar lo, Scalar loReplace, Scalar hi, Scalar h
...
+    return matchesRe(new RegexProgram(pattern, CaptureGroups.NON_CAPTURE));
+  }
+
+  /**
+   * Returns a boolean ColumnVector identifying rows which
+   * match the given regex program pattern but only at the beginning of the string.
+   *
+   * ```
+   * cv = ["abc", "123", "def456"]
+   * p = regex_program::create("\\d+")
```
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ColumnView.java:3178]: These are Javadocs but this is showing examples of C++ code.

Comment (cindyyuanjiang) [reply java/src/main/java/ai/rapids/cudf/ColumnView.java]: Updated. Thank you very much!

**Review by jlowe — APPROVED** [review]

**Review by msadang — APPROVED** [review]
LGTM

**Pull request merged**